inc/azure_c_shared_utility/agenttime.h

```diff
@@ -19,42 +19,46 @@
 extern "C"
 {
 #endif
+//#include <windows.h>
+
+//WINAPI DLLEXPORT int marka(int a, int b);
+//int __declspec(dllexport) __stdcall marka(int a, int b);
 
 /** @brief Get current calendar time.
 *
 *	@details This function provides the same functionality as the
 *	standard C @c time() function.
 */
-MOCKABLE_FUNCTION(, time_t, get_time, time_t*, currentTime);
+DLLEXPORT MOCKABLE_FUNCTION(, time_t, get_time, time_t*, currentTime);
 
 /** @brief Get UTC in @c tm struct.
 *
 *	@details This function provides the same functionality as the
 *	standard C @c gmtime() function.
 */
-MOCKABLE_FUNCTION(, struct tm*, get_gmtime, time_t*, currentTime);
+DLLEXPORT MOCKABLE_FUNCTION(, struct tm*, get_gmtime, time_t*, currentTime);
 
 /** @brief Get current time representation of the given calendar time.
 *
 *	@details This function provides the same functionality as the
 *	standard C @c mktime() function.
 */
-MOCKABLE_FUNCTION(, time_t, get_mktime, struct tm*, cal_time);
+DLLEXPORT MOCKABLE_FUNCTION(, time_t, get_mktime, struct tm*, cal_time);
 
 /** @brief Gets a C-string representation of the given time.
 *
 *	@details This function provides the same functionality as the
 *	standard C @c ctime() function.
 */
-MOCKABLE_FUNCTION(, char*, get_ctime, time_t*, timeToGet);
+DLLEXPORT MOCKABLE_FUNCTION(, char*, get_ctime, time_t*, timeToGet);
```

Comment:
>DLLEXPORT [](start = 0, length = 9)

What does this buy us over .def?

Reply:
UploadToblob flag currently requires 2 .def files -- one with uploadtoblob stuff and the other without.  With modules (sadly) looking like they'll need a separate cmake flag and in turn need custom Dllexports based on whether its built, the combinitorics of managing this with .def files start blowing up.  

We could try to filter out a master .def as we did in WinCE/SYSGEN days but those were bespoke CE tools IIRC.  Building this up based on declspec's seems most natural.

Thoughts?

Reply:
I see.

---
In reply to: [184147428](https://github.com/Azure/azure-c-shared-utility/pull/152#discussion_r184147428) [](ancestors = 184147428)

Reply:
We (me and Andrei) actually have a better (we believe) proposal.

Have one .def file for the most of the SDK.
Have another .def file for the upload to blob part, since that is an optional feature.

If the upload to blob is enabled then put both .def files in the files list in cmake:

set(iothub_client_def_files
    file1.def
    file2.def
)

add_library(iothub_client_dll SHARED
    ${iothub_client_c_files} 
    ${iothub_client_h_files}
    ${iothub_client_def_files}
)

Basically the idea is that DLLEXPORT is special packaging needed for Windows DLL targets.

The source code and packaging should be kept as far as possible from each other.
Or put it another way: the source should have as few knowledge as possible of stuff that is related to packaging.

That keeps code clean and things do not creep in.

HTH.

---
In reply to: [184153567](https://github.com/Azure/azure-c-shared-utility/pull/152#discussion_r184153567) [](ancestors = 184153567,184147428)